Feature/professor lol 22

pci2676 opened the pull request:
1. 네이밍 수정해야함 config의 난무..

2. validation check가 부족함!

3. mock test 하는 코드 보고 이해안되는거 있으면 코멘트 부탁할게!

4. 물론 그 외에도 바꿨으면 하는것들 다 알려줘! 

5. 수정하기전에 그래도 test쪽은 미리 보는게 좋을거 같아서 풀리퀘 날려놓을게!

pci2676 committed:
professor-lol-22 cherry-pick bean config

pci2676 committed:
professor-lol-22 main 리팩토링

프로파일 설정에 따라 RestTemplate 빈 생성되도록 설정

professor-lol-22 test code

match list 가져오는곳 Object 배열로 해야 정상적으로 요청이 가는것을 확인

pci2676 committed:
professor-lol-22 match param 변경

match list 가져오기 parameters -> Map<String, Object>로 변경

pci2676 committed:
professor-lol-22 errorhandler 설정.

error handler 설정
1. client와 server에러로 분기하도록 설정
2. exception dto 생성

pci2676 committed:
professor-lol-22 exception

exception check json_utf_8 인지

pci2676 committed:
professor-lol-22 MatchDto 받아오기

MatchDto의 모든 내용을 받아오도록 작성

pci2676 committed:
professor-lol-22 Test Ignore

Ignore 어노테이션 추가.

professor-lol-22 Test Ignore

Ignore 어노테이션 추가.

professor-lol-22 exception

group exception enum

pci2676 committed:
professor-lol-22 restTempalte config 변경

pci2676 committed:
professor-lol-22 통신 클래스 파라미터 변경.

resttemplate -> resttemplateBuilder

pci2676 committed:
professor-lol-22 mock test / league

professor-lol-22 mock test / match

ignore재설정

professor-lol-22 mock test / 통합테스트

pci2676 committed:
professor-lol-22 stub rest template 작성

pci2676 committed:
professor-lol-22 stub test

professor-lol-22 refactoring

pci2676 committed:
professor-lol-22 profile 전략 설정, 네이밍 수정

1. profile 전략
- major, test, local

2. 네이밍 수정
- config -> profile 에 매치되게 설정

pci2676 committed:
professor-lol-22 validate code 추가

1. exception advice 설정
2. validation test code 추가

pci2676 committed:
professor-lol-22 test code 수정

1. uri 빌드로 중복코드 삭제
2. 솔큐만 검색하도록 수정

chldbtjd2272 commented on professorlol-riot-api/src/main/java/com/css/professorlol/league/impl/LeagueRestTemplateImpl.java:
```diff
@@ -1,49 +1,36 @@
-            throw restClientException;
+    public Set<LeagueEntryDto> getLeagueEntries(final String encryptedSummonerId) {
+        validateSummonerId(encryptedSummonerId);
+        return restTemplate.getForObject(LOL_LEAGUE_V_4_ENTRIES_BY_SUMMONER, LeagueEntrySet.class, encryptedSummonerId);
```
혹시 외부에서 Set으로 안내려줄 경우는 어떻게 동작해?

pci2676 replied:
![image](https://user-images.githubusercontent.com/13347548/64066636-557d3200-cc57-11e9-9ac3-6d96f0cb6d55.png)
외부에서 set으로 안내려주는 경우를 어떻게 확인해야할지 모르겠는데 riot 명세에는 set으로 내려주는 경우에만 대해 명세가 되어있어! 그래서 결과가 없는 경우네는 비어있는 set을 내려줘

csbsjy commented on professorlol-riot-api/src/main/java/com/css/professorlol/match/dto/match/MasteryDto.java:
```diff
@@ -1,14 +1,12 @@
-
-    List<TeamStatsDto> teams;
+public class MasteryDto {
+    Integer masteryId;
```
접근제어자!

pci2676 replied:
땡큐 고쳤어!

pci2676 committed:
professor-lol-22 접근 제어자 수정

1. 접근 제어자가 없던 부분 접근제어자 설정.

csbsjy commented on professorlol-riot-api/src/test/java/com/css/professorlol/match/impl/MatchRestTemplateImplTest.java:
```diff
@@ -1,121 +1,117 @@
-        MatchlistDto matchlist = matchRestTemplate.getMatchList(encryptedAccountId, matchQueryParam);
+        //then exception
+        assertThatThrownBy(() -> matchRestTemplate.getMatchList(encryptedSummonerId, matchQueryParam))
+                .isInstanceOf(BadRequestException.class);
```
민형이가 한얘기랑 비슷한데, 잘못된 종료 시간 값으로 요청하는 경우에는 BadRequestException이고 시작시간으로 요청하는 경우에는 ClientException 으로 설정한 이유가 무엇인가요~~ 둘다 잘못된 입력값인거 같아서요~~

pci2676 replied:
아 그러네 흠...
그런데 일단 이거 잡은 기준은 라이엇에서 던져주는 익셉션을 기준으로 잡았거든,
라이엇에서 배드리퀘스트로 주는것만 내가 에러값만 메세지로 다시 커스텀 배드리퀘스트 익셉션 터트리는거고 나머지는 클라이언트익셉션으로 터트리고있어

csbsjy commented on professorlol-riot-api/src/test/java/com/css/professorlol/match/impl/MatchRestTemplateImplTest.java:
```diff
@@ -1,121 +1,117 @@
-        log.info(gson.toJson(matchQueryParam));
-
-        MultiValueMap param = matchQueryParam.getQueryParam();
+        Long matchId = 3724003831L;
```
이 매치 아이디가 항상 없는 경기인걸 보장하는지 궁금합니다

pci2676 replied:
확인했을때 없는경기라고 나와서 했어!

jyami-kim commented on professorlol-riot-api/src/test/java/com/css/professorlol/MockResponse.java:
```diff
@@ -0,0 +1,2783 @@
+    }
+
+    public static String getSummonerMockBody() {
+        return new JsonParser().parse(
```
이렇게 Mock Json 같은경우는 그냥 test/resource에 .json 파일로 만들어서 불러오는건 이상하려나??
만약 json key-value 값이 변해서 mock json 을 수정해야하는 경우라면, string으로 되있어서 , {} : 뭔가 빼먹은걸 모를 수도 있고,,, test 돌리다가 에러나면, indent 안되있으니까 줄 하나하나 따라가서 봐야할 것 같아서!!

pci2676 replied:
음 근데 Json으로 해도 똑같이 {}이런거 빼먹을수 있는거아니야? json으로 하면 저거 빼먹으면 잘못된 json이라고 알려줘??
저거 mock 스트링 값 riot에 test로 쏜거 긁어서 복붙 한거라 key-value바뀌면 다시 새로운거 복붙 해야한다고 생각했어...

jyami-kim commented on professorlol-riot-api/src/test/java/com/css/professorlol/league/impl/LeagueRestTemplateImplMockTest.java:
```diff
@@ -0,0 +1,129 @@
+@ActiveProfiles("major")
+public class LeagueRestTemplateImplMockTest {
+
+    private static final Logger log = LoggerFactory.getLogger(SummonerRestTemplateImplMockTest.class);
```
여기 혹시 LeagueRestTemplateImpleMockTest.class로 되어야있어야하는거아냐??
file class명이랑 Logger의 class명이랑 달라서!

Gobukgol replied:
클래스명 같아야하는거면 this.getClass() 로 가능할까?

pci2676 replied:
@mjung1798 복붙하다가 못봤나봐! 고쳤어!
@Gobukgol 스태틱한 객체라 this를 쓸수가 없어! static 을 붙이지 말까? 습관적으로 항상 로거를 static으로 붙이고 써서 이렇게 작성한거같아

chldbtjd2272 commented on professorlol-riot-api/src/main/java/com/css/professorlol/config/exception/BadRequestException.java:
```diff
@@ -0,0 +1,7 @@
+package com.css.professorlol.config.exception;
+
+public class BadRequestException extends RiotException {
```
BadRequestException같은 경우는 너무 추상화된 이름인거같아

pci2676 replied:
NotCorrectInputException으로 이름 변경했어!

chldbtjd2272 commented on professorlol-riot-api/src/main/java/com/css/professorlol/config/exception/dto/RiotExceptionDto.java:
```diff
@@ -0,0 +1,31 @@
+@ToString
+@Builder
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor
```
@Builder,@NoArgsConstructor,@AllArgsConstructor 이게 빌더가 되나?
롬복 사용할때 저렇게 붙이면 빌더 안됐던걸로 기억하는데

그리고 @Builder,@AllArgsConstructor는 안쓰는거같은데 뺴도되지않을까

pci2676 replied:
예전에 만들때 이것저것 테스트 하느라 작성하고 안지웠나봐! 제거했어!

chldbtjd2272 commented on professorlol-riot-api/src/main/java/com/css/professorlol/config/exception/handler/RiotErrorHandler.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    private RiotExceptionDto parseRiotExceptionBody(ClientHttpResponse response) throws IOException {
+        if (isJson(response)) {
```
제이슨 인지 검사안하고 그냥 try catch로 감싸면 될듯

pci2676 replied:
변경했어!

chldbtjd2272 commented on professorlol-riot-api/src/main/java/com/css/professorlol/config/exception/handler/RiotErrorHandler.java:
```diff
@@ -0,0 +1,72 @@
+
+    private RiotExceptionDto parseRiotExceptionBody(ClientHttpResponse response) throws IOException {
+        if (isJson(response)) {
+            InputStreamReader inputStreamReader = new InputStreamReader(response.getBody());
```
String exceptionBody = String.valueOf(response.getBody());

inputStream을 그냥 String으로 바꾸면 안되는거야??

pci2676 replied:
그렇게 하니까 토큰을 못찾는 에러가 발생하더라고 그래서 저렇게 파싱했어

pci2676 replied:
이게 getBody()가 InputStream을 반환해줘서 저런방식으로 파싱해야되는 것 같아

chldbtjd2272 commented on professorlol-riot-api/src/main/java/com/css/professorlol/config/exception/handler/RiotErrorHandler.java:
```diff
@@ -0,0 +1,72 @@
+        throw new NotJsonTypeException(response.getStatusText());
+    }
+
+    private boolean isJson(ClientHttpResponse response) {
```
try catch로 잡으면 이거 없어도 될듯

chldbtjd2272 commented on professorlol-riot-api/src/main/java/com/css/professorlol/config/resttemplate/LeagueRestTemplateConfig.java:
```diff
@@ -0,0 +1,53 @@
+    @Configuration
+    public static class MajorLeagueConfig {
+
+        private final Duration ONE_SEC = Duration.ofMillis(1000);
```
private final은 빈주입 대상이라
static 붙여주는게 좋을듯

chldbtjd2272 commented on professorlol-riot-api/src/main/java/com/css/professorlol/config/exception/handler/RiotErrorHandler.java:
```diff
@@ -0,0 +1,72 @@
+    @Override
+    public void handleError(ClientHttpResponse response) throws IOException {
+        RiotExceptionDto exceptionDto = parseRiotExceptionBody(response);
+        if (isBadRequest(response)) {
```
dto로 검증하는게 맞을듯

chldbtjd2272 commented on professorlol-riot-api/src/main/java/com/css/professorlol/config/exception/handler/RiotErrorHandler.java:
```diff
@@ -0,0 +1,72 @@
+        throw new ServerException(exceptionDto.getMessage());
+    }
+
+    private boolean isClientError(ClientHttpResponse response) throws IOException {
```
요런 간단한 로직은 dto안에 넣는게 좋을거같아

pci2676 replied:
근데 hasError에서 검증하려면 ErrorHandler가 들고 있는게 맞지않아?
handleError에서 dto가 자신이 들고있는 error가 400인지만 확인해 주면 되는거아니야?
그래서 일단 400인지 확인하는 로직만 내부로 옮겼어!

chldbtjd2272 commented on professorlol-riot-api/src/main/java/com/css/professorlol/summoner/impl/SummonerRestTemplateImpl.java:
```diff
@@ -1,47 +1,41 @@
+    }
+
+    private void validateSummonerIdentifier(final String summonerIdentifier) {
+        if (summonerIdentifier == null || summonerIdentifier.isEmpty()) {
```
StringUtils.isEmpty()로 한번에 검사할 수 있어

pci2676 replied:
변경했어 땡큐

chldbtjd2272 commented on professorlol-riot-api/src/test/java/com/css/professorlol/league/impl/LeagueRestTemplateImplMockTest.java:
```diff
@@ -0,0 +1,129 @@
+    private MockRestServiceServer mockServer;
+
+    @Autowired
+    private LeagueRestTemplate leagueRestTemplate;
```
이걸 꼭 빈으로 띄워서 테스트 하지 않는 방법도 있어
mockServer.createServer();를 이용해서 이렇게 하면 스프링 의존성 없이 테스트를 짤 수 있어서 더 좋은 테스트가 될거같아

pci2676 replied:
![image](https://user-images.githubusercontent.com/13347548/64066844-e81ed080-cc59-11e9-988a-885b95b4e573.png)
음 이렇게 하는거라고 생각해서 해봤는데 이러니까 restTemplate가 mocking되지 않는거 같아

chldbtjd2272 commented on professorlol-riot-api/src/main/java/com/css/professorlol/config/exception/handler/RiotErrorHandler.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.Optional;
+
+@Slf4j
+public class RiotErrorHandler implements ResponseErrorHandler {
```
이건 테스트 코드가 있어야 할 부분인거같은데 테스트가 없네

ClientHttpResponse같은 경우에는 MockClientHttpResponse로 생성해서 테스트를 할 수 있어

참고:https://www.programcreek.com/java-api-examples/index.php?api=org.springframework.mock.http.client.MockClientHttpResponse

pci2676 replied:
테스트 작성했으! 참고땡큐!

chldbtjd2272 commented on professorlol-riot-api/src/test/java/com/css/professorlol/summoner/impl/SummonerRestTemplateImplMockTest.java:
```diff
@@ -0,0 +1,184 @@
+        SummonerDto summonerDto = this.summonerRestTemplate.getSummonerDtoBySummonerId(summonerId);
+
+        //then
+        assertThat(summonerDto).isNotNull();
```
notNull의 검증은 의미가 없는듯

pci2676 replied:
summonerId를 검증하도록 변경했어!

pci2676 committed:
professor-lol-22 host url 위치 변경

host url은 x riot properties가 들고 있기에는 성격이 맞는것 같지 않아서
빌더가 있는 곳으로 뺐음

Gobukgol commented on professorlol-riot-api/src/main/java/com/css/professorlol/summoner/impl/SummonerRestTemplateImpl.java:
```diff
@@ -1,47 +1,41 @@
+    }
+
+    @Override
+    public SummonerDto getSummonerDto(final String summonerName) {
```
소환사명으로 dto 가져온다는걸 함수명에 명시해주면 좋을것 같아

pci2676 replied:
전부 일관성있게 바꿨어!

pci2676 committed:
professor-lol-22 static 추가

bean 주입대상이 되지않도록 변경

Gobukgol commented on professorlol-riot-api/src/test/java/com/css/professorlol/league/impl/LeagueRestTemplateImplTest.java:
```diff
@@ -1,42 +1,83 @@
+        log.info(gson.toJson(leagueEntryDtoSet));
+    }
+
+    @Test(expected = BadRequestException.class)
```
밑에 ThrownBy() 있으니까 없어도 가능하지않아?

pci2676 replied:
변경하다가 못보고 지나쳤나봐 고마워 지웠어!

pci2676 committed:
professor-lol-22 Exception name change

BadRequestException 이름을 NotCorrectInputException 으로 변경

professor-lol-22 불필요한 롬복 어노테이션 제거

RiotExceptionDto 에 불필요한 어노테이션 제거

professor-lol-22 BadRequest check

400 에러인지 dto 에서 체크하도록 변경

professor-lol-22 validation 간소화

StringUtils.isEmpty() 를 이용하여 간소화

professor-lol-22 불필요한 test 어노테이션 제거

expect 에 있던 불필요한 exception test 어노테이션 정리

professor-lol-22 logger class 변경

잘못 작성된 클래스명 변경

professor-lol-22 테스트 코드 변경

professor-lol-22 json parsing 코드 변경

try-catch 로 exception 발생

professor-lol-22 불필요한 로그 제거

professor-lol-22 메소드 이름 수정

어떤 파라미터로 가져오는지 명시적으로 수정

professor-lol-22 빌더 접근제한자 변경

private 이었는데 외부에서 시용하려면 public 이맞는거같아

professor-lol-22 error handler 테스트 추가

error handler 테스트 추가

professor-lol-22 리뷰 내용 반영

1. error handler 변경
2. exception dto 변경
3. mock string -> .json 변경

Gobukgol commented on professorlol-riot-api/src/main/java/com/css/professorlol/config/exception/ClientException.java:
```diff
@@ -0,0 +1,7 @@
+package com.css.professorlol.config.exception;
+
+public class ClientException extends RiotException {
```
ClientException의 의미가 모호한거 같아. 밑 테스트 코드 중에서 잘못된 시간 값을 입력했을 때에 ClientException을 주는데 결국엔 이것도 BadRequest에 포함되는것 같아.

pci2676 replied:
이게 라이엇에서 잘못된 시간값 입력한걸 BadRequest로 내려주지않아서 ClientException으로 넘어갔어!
이거 그러면 시간값에 대한 validation 체크는 요청하기 전에해서 다른 Exception을 발생시키도록 변경해야 할것 같아!
